x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java

```diff
@@ -197,8 +206,65 @@ private void innerExecute(Task task, EsqlQueryRequest request, ActionListener<Es
             remoteClusterService,
             planRunner,
             queryBuilderResolver,
-            listener.map(result -> toResponse(task, request, configuration, result))
+            ActionListener.wrap(result -> {
+                recordCCSTelemetry(task, executionInfo, request, null);
+                listener.onResponse(toResponse(task, request, configuration, result));
+            }, ex -> {
+                recordCCSTelemetry(task, executionInfo, request, ex);
+                listener.onFailure(ex);
+            })
         );
+
+    }
+
+    private void recordCCSTelemetry(Task task, EsqlExecutionInfo executionInfo, EsqlQueryRequest request, @Nullable Exception exception) {
+        if (executionInfo.isCrossClusterSearch() == false) {
+            return;
+        }
+
+        CCSUsage.Builder usageBuilder = new CCSUsage.Builder();
+        usageBuilder.setClientFromTask(task);
+        if (exception != null) {
+            if (exception instanceof VerificationException) {
```

Comment:
Nit: if you rewrite this as `exception instanceof VerificationException v`, then you can eliminate the casting in the below line and pass `v` directly to `classifyVerificationException()`.